pkg/router/base_rule.go

```diff
@@ -202,6 +205,12 @@ func (rri *RouteRuleImplBase) finalizeRequestHeaders(headers api.HeaderMap, requ
 	rri.vHost.globalRouteConfig.requestHeadersParser.evaluateHeaders(headers, requestInfo)
 	if len(rri.hostRewrite) > 0 {
 		headers.Set(protocol.IstioHeaderHostKey, rri.hostRewrite)
+	} else if rri.autoHostRewrite && configmanager.IsDnsTypeCluster(rri.ClusterName()) {
```

Comment:
`autoHostRewrite` means that the host header will be swapped with the hostname of the upstream host chosen by the cluster manager and not `ClusterName`.